### PR TITLE
CASSANDRA-18155 Coordinator level metrics for read response and mutation row and column counts

### PR DESCRIPTION
**Pull request by maedhroz**

### CODE REVIEW

**Comment by maedhroz** [comment]
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-18155

File: src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java [hunk 1]
```diff
@@ -228,5 +228,9 @@ private boolean handleInFilter(SingleRestriction restriction, int index)
     {
         return restriction.isContains() || restriction.isLIKE() || index != restriction.getFirstColumn().position();
     }
-
+    
+    public RestrictionSet getRestrictionSet()
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java]: Could this be more useful in `RestrictionSetWrapper` ? At the end of the day it's like calling `getColumnDefs()` and iterating while calling `getRestrictions(ColumnMetadata columnDef)`, which all live in the super.

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java [hunk 2]
```diff
@@ -104,6 +104,17 @@ public List<ColumnMetadata> getColumnDefs()
     {
         return new ArrayList<>(restrictions.keySet());
     }
+    
+    @Override
+    public Collection<ColumnMetadata> getColumnDefinitions()
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:112]: Why not use the already existing `getColumnDefs()`?

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java [hunk 3]
```diff
@@ -104,6 +104,17 @@ public List<ColumnMetadata> getColumnDefs()
...
+    {
+        return restrictions.keySet();
+    }
+
+    public Map<ColumnMetadata, SingleRestriction> getRestrictions()
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java]: We're already iterable do we need this?

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSetWrapper.java [hunk 4]
```diff
@@ -57,6 +58,12 @@ public List<ColumnMetadata> getColumnDefs()
         return restrictions.getColumnDefs();
     }
 
+    @Override
+    public Collection<ColumnMetadata> getColumnDefinitions()
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/RestrictionSetWrapper.java:62]: Same, am I missing sthg?

File: src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java [hunk 5]
```diff
@@ -390,6 +390,53 @@ public boolean isColumnRestrictedByEq(ColumnMetadata columnDef)
...
+    /**
+     * This method determines whether a specified column is restricted on equality or something equivalent, like IN.
+     * It can be used in conjunction with the columns selected by a query to determine which of those columns is 
+     * already bound by the client and not necessarily retrieved by the database.
+     *
+     * @param column a column specification from the same table these restrictions are against
+     *
+     * @return true if the given column is restricted on equality
+     */
+    public boolean isEqualityRestricted(ColumnSpecification column)
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java]: Here you're potentially traversing all columns on each call. Given you're on a hot path, could you be feeding instead this method with `ColumnMetada`, pulled from the restrictions, and relaying on `KIND` to spare a couple expensive loops?

File: src/java/org/apache/cassandra/db/partitions/PartitionUpdate.java [hunk 6]
```diff
@@ -466,6 +466,56 @@ public List<CounterMark> collectCounterMarks()
...
+    {
+        // If there is a partition-level deletion, we intend to delete at least the columns of one row.
+        if (!partitionLevelDeletion().isLive())
+            return metadata().regularAndStaticColumns().size();
+
+        int count = 0;
+
+        // Each range delete should correspond to at least one intended row deletion, and with it, its regular columns.
+        if (deletionInfo().hasRanges())
+            count += deletionInfo().rangeCount() * metadata().regularColumns().size();
```
Comment (bereng) [inline src/java/org/apache/cassandra/db/partitions/PartitionUpdate.java:507]: +stattics? +counters?

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 3 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java]: It actually looks like this is simply unused. Removing it...

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 2 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:112]: I think the worry there (for performance) was what appeared to be a needless defensive copy in `getColumnDefs()`. I didn't want to audit the 12 other existing callers of `getColumnDefs()`. Certainly open to naming changes...

File: src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java (hunk 1 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java]: Makes sense...I'll pull it up.

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSetWrapper.java (hunk 4 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSetWrapper.java:62]: see above (tl;dr avoid defensive copy)

File: src/java/org/apache/cassandra/db/partitions/PartitionUpdate.java (hunk 6 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/db/partitions/PartitionUpdate.java:507]: I'm not 100% sure on counters. They are just regular columns, correct? (Going up a level, this is really a pretty conservative lower bound estimate, given range deletes can delete more than one row. There just isn't a way to figure that out without actually doing a read, etc.)

Good catch on the statics. They wouldn't be covered by range deletes, but I'm still not accounting for them properly in either `affectedRowCount()` or `affectedColumnCount()`. Working on a couple new tests, and should have something up shortly...

File: src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java (hunk 5 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java]: Let me mull this one over. In the meantime though, note that we can disable all of this w/ `client_request_size_metrics_enabled` in the YAML.

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 2 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:109]: Also, we may be able to delete these entirely, depending on what we do w/ https://github.com/apache/cassandra/pull/2105#discussion_r1083887949

File: src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java (hunk 5 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java]: You can disable it yes, but when enabled you want it as performant as possible imo.

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 2 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:112]: Ohhh icwym with the defensive copy. Then I would either document it, rename it or add a flag to choose the desired behavior. Anything that makes it obvious why there are 2 options to choose.

**Comment by maedhroz** [comment]
@bereng I've pushed changes to properly account for statics on the write side and make `isEqualityRestricted()` less expensive. Let me know what you think...

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 2 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:112]: Agreed. Will at least JavaDoc that for posterity...

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java [hunk 7]
```diff
@@ -105,6 +105,15 @@ public List<ColumnMetadata> getColumnDefs()
         return new ArrayList<>(restrictions.keySet());
     }
 
+    /**
+     * @return a direct reference to the key set from {@link #restrictions} with no defenseive copying
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:109]: nit: `a` -> `A direct reference...`

File: src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java [hunk 8]
```diff
@@ -390,6 +390,54 @@ public boolean isColumnRestrictedByEq(ColumnMetadata columnDef)
...
+                    if (restricted.name.equals(column.name))
+                        return true;
+        }
+        else if (column.kind == ColumnMetadata.Kind.CLUSTERING)
+        {
+            if (hasClusteringColumnsRestrictions())
+            {
+                for (SingleRestriction restriction : clusteringColumnsRestrictions.getRestrictionSet())
+                {
+                    if (restriction.isEQ() || restriction.isIN())
```
Comment (bereng) [inline src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java]: Did we forget about `CONTAINS` here?

Comment (bereng) [reply src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java]: Maybe you want to collect the result in a variable and return at the end. That might help on branch prediction by avoiding early returns, although iirc most modern cpus are already optimized on that front.

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 7 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:109]: see https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#format

tl;dr I think the spirit of the official formatting style there is that the `@return` is sort of like the first word in a "sentence"

File: src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java (hunk 8 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java]: In the sense that `CONTAINS` and `CONTAINS_KEY` both deal w/ information provided by the user, I think it makes sense to include them if we're looking at whether or not the metric counts them as "read". I'll throw together a couple tests. While I'm at it, something like `isEqualityBased()` on `SingleRestriction` to avoid a messy/long if condition here probably also makes sense.

File: src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java (hunk 7 above)
Comment (bereng) [reply src/java/org/apache/cassandra/cql3/restrictions/RestrictionSet.java:109]: TIL. Works for me.

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/4892331eb49f72d2e18432a2af56538d11c7c5d1
